Allow the user to create a Client from a PacketConn

paultag opened the pull request:
In particular, on Linux, this will allow the user to create an
arp.Client in promiscious mode, by having raw.ListenPacket call socket(2)
with htons(ETH_P_ALL) by passing `syscall.ETH_P_ALL` into it.

An example replacement of an existing call site would look something
like:

```
dev, err := net.InterfaceByName("name0")
if err != nil {
    return err
}
packetConn, err := raw.ListenPacket(dev, syscall.ETH_P_ALL)
if err != nil {
    return err
}
arpClient, err := arp.NewClientPacketConn(dev, packetConn)
...
```

paultag commented:
Also worth noting, I have no idea where the correct interface point would be - I just figured I'd try to make a minimal change that won't break API, and will work on the arches supported. I'm more than happy to move this somewhere more sensible

shurcooL commented on client.go:
```diff
@@ -33,7 +33,10 @@ func NewClient(ifi *net.Interface) (*Client, error) {
+	return NewClientPacketConn(ifi, p)
+}
 
+func NewClientPacketConn(ifi *net.Interface, p net.PacketConn) (*Client, error) {
```
Go style suggests that all top-level, exported names should have doc comments.

Reference: https://github.com/golang/go/wiki/CodeReviewComments#doc-comments.

Also, Go lint would say:

```
exported function NewClientPacketConn should have comment or be unexported
```

paultag replied:
I was sending this up for feedback on where to extend the API, and I wasn't keen to document a method I was unsure of. However, since this is the only feedback, I'm going to assume the content is acceptable, and I've pushed docs.

shurcooL replied:
I see. My feedback was about a generic Go style issue I noticed. I think only the project owner will be able to tell whether this change makes sense/belongs here at a higher level.

paultag replied:
Oh, I thought because you were reviewing PRs, you were a contributor. Oh well. At most I only wasted a small amount of time -- I probably would have held off until I was sure this was going to be accepted.

shurcooL commented on client.go:
```diff
@@ -36,6 +36,11 @@ func NewClient(ifi *net.Interface) (*Client, error) {
+// and raw socket. This allows the caller to define exactly how they bind to the
+// raw socket. This is most useful to define what protocol to pass to socket(7).
+//
+// In most cases, callers would be better off calling NewClient
```
Nitpick, this sentence is missing a period at the end.

paultag replied:
(Added)

mdlayher commented on client.go:
```diff
@@ -33,7 +33,15 @@ func NewClient(ifi *net.Interface) (*Client, error) {
+}
 
+// NewClientPacketConn creates a new Client using the specified network interface
+// and raw socket. This allows the caller to define exactly how they bind to the
```
Instead of "raw socket", how about "any net.PacketConn" for accuracy?

mdlayher commented on client.go:
```diff
@@ -33,7 +33,15 @@ func NewClient(ifi *net.Interface) (*Client, error) {
 
+// NewClientPacketConn creates a new Client using the specified network interface
+// and raw socket. This allows the caller to define exactly how they bind to the
+// raw socket. This is most useful to define what protocol to pass to socket(7).
```
Replace "raw socket" with "net.PacketConn".

paultag commented:
@mdlayher Yeah, cleaning up API to not have such a clunky name (I'm not a fan of how `NewClientPacketConn` reads either) would be good -- happy to help with that -- I've fixed the bits in the docstring, and squashed the changes into one commit
